dispatches/case_studies/renewables_case/RE_surrogate_optimization_steadystate.py

```diff
@@ -0,0 +1,341 @@
+    
+    m.pem_system_capacity = Var(domain=NonNegativeReals, bounds=(127.05 * 1e3, 423.5 * 1e3), initialize=input_params['pem_mw'] * 1e3, units=pyunits.kW)
+    m.pem_bid = Var(within=NonNegativeReals, bounds=(15, 45), initialize=45)                    # Energy Bid $/MWh
+    m.reserve_percent = Param(within=NonNegativeReals, initialize=15)   # Reserves Fraction on Grid
+    m.shortfall_price = Param(within=NonNegativeReals, initialize=1000)     # Energy price during load shed
+
+    inputs = [m.pem_bid, m.pem_system_capacity * 1e-3 / 847 * m.wind_system_capacity * 1e-3, m.reserve_percent, m.shortfall_price]
+
+    # extant_wind means if it is a built plant. If true, the captial cost is 0.
+    if input_params['extant_wind']:
+        m.wind_system_capacity.fix()
+    if PEM_bid is not None:
+        m.pem_bid.fix(PEM_bid)
+    if PEM_MW is not None:
+        m.pem_system_capacity.fix(PEM_MW * 1e3)
+
+    # add NN surrogates to the model using omlt
+    ##############################
+    # revenue surrogate
+    ##############################
+    m.rev_surrogate = Var()
+    m.nn_rev = omlt.OmltBlock()
+    formulation_rev = FullSpaceNNFormulation(net_rev_defn)
+    m.nn_rev.build_formulation(formulation_rev)
+    
+    m.constraint_list_rev = ConstraintList()
+    for i in range(len(inputs)):
+        m.constraint_list_rev.add(inputs[i] == m.nn_rev.inputs[i])
+    m.constraint_list_rev.add(m.rev_surrogate == m.nn_rev.outputs[0])
+
+    # make rev non-negative, MM$
+    # m.rev = Expression(expr=0.5 * pyo.sqrt(m.rev_surrogate**2 + 0.001**2) + 0.5 * m.rev_surrogate)
+    m.rev = Expression(expr=m.rev_surrogate)
+
+    ##############################
+    # dispatch frequency surrogate
+    ##############################
+    m.dis_set = RangeSet(0, num_rep_days - 1)
+    m.frequency_surrogate = Var(m.dis_set, initialize=[(x + 1) / (x + 1) / num_rep_days for x in range(num_rep_days)])
+    m.nn_frequency = omlt.OmltBlock()
+    formulation_frequency = FullSpaceNNFormulation(net_frequency_defn)
+    m.nn_frequency.build_formulation(formulation_frequency)
+
+    m.constraint_list_dispatch = ConstraintList()
+    for i in range(len(inputs)):
+        m.constraint_list_dispatch.add(inputs[i] == m.nn_frequency.inputs[i])
+
+    m.frequency_surrogate_nonneg = Var(m.dis_set, initialize=[(x + 1) / (x + 1) / num_rep_days for x in range(num_rep_days)])
+    for i in range(num_rep_days):
+        m.constraint_list_dispatch.add(m.frequency_surrogate_nonneg[i] == 0.5 * pyo.sqrt(m.nn_frequency.outputs[i]**2 + 0.001**2) + 0.5 * m.nn_frequency.outputs[i])
+    
+    # sum of outputs of dispatch surrogate may not be 1, scale them. 
+    for i in range(num_rep_days):
+        m.constraint_list_dispatch.add(m.frequency_surrogate[i] == m.frequency_surrogate_nonneg[i] / sum(value(m.frequency_surrogate_nonneg[j]) for j in range(num_rep_days)))
```

Comment:
Any reason to use `ConstraintList` here?

Reply:
Not sure, this code was written by Xian